### ISSUE TITLE
[MaterialCardView] CardView will not shrink after first measure

### ISSUE THREAD

**Issue by rvenk** [material-components/material-components-android#537]
**Description:** 
We tried to use a clickable card view in a `GridLayout`, but on compat devices the card views are not scaled properly, this is because the underlying `FrameLayout` takes the foreground drawable into account when measuring.

The current behavior is as follows.
1. We have a `GridLayout` with clickable card views, the card views have a width of 0 dp to distribute the space evenly.
2. In the first measure, the card view is really high because the textview in the card view don't have enough space.
3. After the first measure, the card view helper sets the inset of the `clickableForegroundDrawable` checked-icon layer to match the measured width and height.
4. The grid layout measures all views again because it now knows the exact width each column should be. The textviews in the card view now have enough width to fit inside the card in a single line.
5. In the second `onMeasure`, the `CardView` now takes the foreground drawable into account, which has a minimum height due to the insets of the checked-icon layer. 
Due to this, the card view can never shrink.

**Expected behavior:** 
In onMeasure, the CardView should not take the `clickableForegroundDrawable` into account.

**Source code:** 
```
<androidx.gridlayout.widget.GridLayout
            android:layout_width="match_parent"
            android:layout_height="wrap_content">

            <com.google.android.material.card.MaterialCardView
                style="@style/Widget.MaterialComponents.CardView"
                android:layout_width="0dp"
                android:layout_height="wrap_content"
                android:clickable="true"
                android:focusable="true"
                app:layout_columnWeight="1"
                app:layout_gravity="fill_vertical|fill_horizontal">

                <androidx.constraintlayout.widget.ConstraintLayout
                    android:layout_width="match_parent"
                    android:layout_height="wrap_content"
                    android:paddingLeft="@dimen/view_margin_lg"
                    android:paddingTop="@dimen/view_margin_lg"
                    android:paddingRight="@dimen/view_margin_lg"
                    android:paddingBottom="@dimen/view_margin_lg">

                    <ImageView
                        android:id="@+id/getInfo_icon"
                        iconColor="@{getInfoHeader.getTextColors()}"
                        iconDrawable="@{FontAwesome.Icon.info_circle}"
                        iconSize="@{@dimen/icon_size_lg}"
                        android:layout_width="wrap_content"
                        android:layout_height="wrap_content"
                        android:contentDescription="@string/no_content_description"
                        android:gravity="center"
                        app:layout_constraintBottom_toBottomOf="parent"
                        app:layout_constraintStart_toStartOf="parent"
                        app:layout_constraintTop_toTopOf="parent" />

                    <TextView
                        android:id="@+id/getInfo_header"
                        android:layout_width="wrap_content"
                        android:layout_height="wrap_content"
                        android:layout_marginStart="@dimen/view_margin_lg"
                        android:layout_marginLeft="@dimen/view_margin_lg"
                        android:text="@string/main_getInfo_title"
                        android:textAppearance="@style/TextAppearance.MaterialComponents.Headline5"
                        app:layout_constrainedWidth="true"
                        app:layout_constraintEnd_toEndOf="parent"
                        app:layout_constraintHorizontal_bias="0.0"
                        app:layout_constraintStart_toEndOf="@+id/getInfo_icon"
                        app:layout_constraintTop_toTopOf="parent" />

                    <TextView
                        android:id="@+id/getInfo_description"
                        android:layout_width="wrap_content"
                        android:layout_height="wrap_content"
                        android:layout_marginStart="@dimen/view_margin_lg"
                        android:layout_marginLeft="@dimen/view_margin_lg"
                        android:hint="@string/main_getInfo_description"
                        android:textAppearance="@style/TextAppearance.MaterialComponents.Body2"
                        app:layout_constrainedWidth="true"
                        app:layout_constraintEnd_toEndOf="parent"
                        app:layout_constraintHorizontal_bias="0.0"
                        app:layout_constraintStart_toEndOf="@+id/getInfo_icon"
                        app:layout_constraintTop_toBottomOf="@id/getInfo_header" />

                </androidx.constraintlayout.widget.ConstraintLayout>
            </com.google.android.material.card.MaterialCardView>

            <com.google.android.material.card.MaterialCardView
                style="@style/Widget.MaterialComponents.CardView"
                android:layout_width="0dp"
                android:layout_height="wrap_content"
                android:clickable="true"
                android:focusable="true"
                app:layout_columnWeight="1"
                app:layout_gravity="fill_vertical|fill_horizontal">

                <androidx.constraintlayout.widget.ConstraintLayout
                    android:layout_width="match_parent"
                    android:layout_height="wrap_content"
                    android:paddingLeft="@dimen/view_margin_lg"
                    android:paddingTop="@dimen/view_margin_lg"
                    android:paddingRight="@dimen/view_margin_lg"
                    android:paddingBottom="@dimen/view_margin_lg">

                    <ImageView
                        android:id="@+id/performAction_icon"
                        iconColor="@{performActionHeader.getTextColors()}"
                        iconDrawable="@{FontAwesome.Icon.arrow_circle_right}"
                        iconSize="@{@dimen/icon_size_lg}"
                        android:layout_width="wrap_content"
                        android:layout_height="wrap_content"
                        android:contentDescription="@string/no_content_description"
                        android:gravity="center"
                        app:layout_constraintBottom_toBottomOf="parent"
                        app:layout_constraintStart_toStartOf="parent"
                        app:layout_constraintTop_toTopOf="parent" />

                    <TextView
                        android:id="@+id/performAction_header"
                        android:layout_width="wrap_content"
                        android:layout_height="wrap_content"
                        android:layout_marginStart="@dimen/view_margin_lg"
                        android:layout_marginLeft="@dimen/view_margin_lg"
                        android:text="@string/main_performAction_title"
                        android:textAppearance="@style/TextAppearance.MaterialComponents.Headline5"
                        app:layout_constrainedWidth="true"
                        app:layout_constraintEnd_toEndOf="parent"
                        app:layout_constraintHorizontal_bias="0.0"
                        app:layout_constraintStart_toEndOf="@+id/performAction_icon"
                        app:layout_constraintTop_toTopOf="parent" />

                    <TextView
                        android:id="@+id/performAction_description"
                        android:layout_width="wrap_content"
                        android:layout_height="wrap_content"
                        android:layout_marginStart="@dimen/view_margin_lg"
                        android:layout_marginLeft="@dimen/view_margin_lg"
                        android:hint="@string/main_performAction_description"
                        android:textAppearance="@style/TextAppearance.MaterialComponents.Body2"
                        app:layout_constrainedWidth="true"
                        app:layout_constraintEnd_toEndOf="parent"
                        app:layout_constraintHorizontal_bias="0.0"
                        app:layout_constraintStart_toEndOf="@+id/performAction_icon"
                        app:layout_constraintTop_toBottomOf="@id/performAction_header" />

                </androidx.constraintlayout.widget.ConstraintLayout>
            </com.google.android.material.card.MaterialCardView>
</androidx.gridlayout.widget.GridLayout>
```

If I set the foreground drawable of the card views to `null` after `onCreate`, the cards get the correct height.

**Android API version:**
Tested on level 19 and 23 and occurs on both devices.

**Material Library version:** 
1.1.0-alpha09

**Device:** 
Motorola TC-55 and Samsung SM-903F

**Comment by rvenk** [comment]
Issue can be quickfixed by creating a drawable wrapper that sets the minimum width / height to -1.
```
public class CardViewForegroundFix extends LayerDrawable {
    public CardViewForegroundFix(Drawable foreground) {
        super(new Drawable[]{foreground});
    }

    @Override
    public int getMinimumHeight() {
        return -1;
    }

    @Override
    public int getMinimumWidth() {
        return -1;
    }

    public static void fix(MaterialCardView cardView) {
        Drawable foreGround = cardView.getForeground();
        if (!(foreGround instanceof CardViewForegroundFix)) {
            final CardViewForegroundFix foregroundFix = new CardViewForegroundFix(foreGround);
            cardView.setForeground(foregroundFix);
            // foreGround#callback is nullified in cardView#setForeground,
            // reset it to the foreground fix drawable
            foreGround.setCallback(foregroundFix);
        }
    }
}
```

If I'm not mistaken, this issue can be fixed in the lib by letting the `InsetDrawable` returned from `CardView#insetDrawable` also have no minimum width / height:
```
// MaterialCardViewHelper.java#529
    return new InsetDrawable(
        originalDrawable, insetHorizontal, insetVertical, insetHorizontal, insetVertical) {
      @Override
      public boolean getPadding(Rect padding) {
        // Our very own special InsetDrawable that pretends it does not have padding so that
        // using it as the background will *not* change the padding of the view.
        return false;
      }

      // *** Fix start ***
      @Override
      public int getMinimumHeight() {
          return -1;
      }

      @Override
      public int getMinimumWidth() {
          return -1;
      }
      // *** Fix end ***
    };
```

**Comment by ymarian** [comment]
Thanks a lot I confirmed your fix and workaround make sense. Will add them.

**Comment by mzroth** [comment]
I see that the fix has been implemented in version 1.2.0-alpha01 but not in 1.1.0-beta01. Is there a timeframe on this being added to the 1.1.0 beta?

Thanks
